assembly/runtime.ts

```diff
@@ -677,37 +669,54 @@ export abstract class BaseContext {
   onDone_: (thiz: BaseContext) => bool;
   onDelete_: (thiz: BaseContext) => void;
 
-  constructor() {
+  constructor(context_id_: u32) {
+    this.context_id = context_id_;
     this.onDone_ = (thiz: BaseContext) => { return thiz.onDone(); };
     this.onDelete_ = (thiz: BaseContext) => { thiz.onDelete(); };
   }
 
   /**
    * Sets the effective context id to this context. this is useful for example if you receive an
    * http call in a RootContext, and want to modify headers based on the response in a regular 
-   * Context. You then will call `context.setEffectiveContext()` so that the header manipulation will
+   * Context. You then will call `setEffectiveContext(_id)` so that the header manipulation will
    * occur in the requerst context and not in the root context.
-   * the r
-   * @param effective_context_id 
+   * @param _id 
    */
-  setEffectiveContext(): WasmResultValues {
-    return imports.proxy_set_effective_context(this.context_id);
+  setEffectiveContext(_id: u32): WasmResultValues {
+    const result = imports.proxy_set_effective_context(_id);
+    if (result != WasmResultValues.Ok) {
+      log(LogLevelValues.critical, "Unable to set effective context: " + _id.toString() + " with result: "+ result.toString());
+    }
+    return result;
+  }
+
+  continueRequest(): void {
+    const result = imports.proxy_continue_request();
+    if (result != WasmResultValues.Ok) {
+      log(LogLevelValues.critical, "Unable to continue request: " + result.toString());
+    }
   }
 
-  // abstract createContext(context_id:u32):Context;
-  onDone(): bool { return true; } // Called when the VM is being torn down.
-  onDelete(): void { } // Called when the VM is being torn down.
+  // Called when the VM is being torn down.
+  onDone(): bool { 
+    log(LogLevelValues.debug, "context id: " + this.context_id.toString() + ": onDone()");
+    return true; 
+  }
+
+  // Called when the VM is being torn down.
+  onDelete(): void { 
+    log(LogLevelValues.debug, "context id: " + this.context_id.toString() + ": onDelete()");
+  } 
 }
 
 /**
- * Wrapper around http callbacks. when asm script supports callbacks, we can refactor \ remove this.
+ * Wrapper around http callbacks. When AS script supports closures, we can refactor \ remove this.
  */
-class HttpCallback {
-  ctx: Object;
-  cb: (c: Object) => void;
-  call(): void { this.cb(this.ctx); }
-  constructor(ctx: Object, cb: (c: Context) => void) {
-    this.ctx = ctx;
+export class HttpCallback {
+  origin_context_id: u32;
+  cb: (origin_context_id: u32, headers: u32,  body_size: usize, trailers: u32) => void;
+  constructor(origin_context_id: u32, cb: (origin_context_id: u32, headers: u32,  body_size: usize, trailers: u32) => void) {
+    this.origin_context_id = origin_context_id;
```

Comment:
This is the way I found to pass along the context_id of the context that is making the http_request. As you mentioned on your comments, we can modify this to use closures once that is available on AssemblyScript.